// @sudoku/stores/game-session.js
import { writable } from 'svelte/store';
import { decodeSencode } from '@sudoku/sencode';
import { generateSudoku, solveSudoku } from '@sudoku/sudoku';
import { createGame, createSudoku } from '../../../domain/index.js';

function createEmptyGrid() {
	return Array.from({ length: 9 }, () => Array(9).fill(0));
}

function toInvalidCellKeys(conflicts) {
	return conflicts.map(({ row, col }) => `${col},${row}`);
}

function snapshotGame(game) {
	return {
		game,
		initialGrid: game.getInitialGrid(),
		currentGrid: game.getSudoku().getGrid(),
		invalidCells: toInvalidCellKeys(game.getConflictingCells()),
		solved: game.isSolved(),
		canUndo: game.canUndo(),
		canRedo: game.canRedo(),
	};
}

function createGameSession() {
	const seedGame = createGame({ sudoku: createSudoku(createEmptyGrid()) });
	const store = writable(snapshotGame(seedGame));
	const { subscribe, set, update } = store;

	const mutateGame = (mutator) => {
		let changed = false;
		update((state) => {
			changed = mutator(state.game);
			return snapshotGame(state.game);
		});
		return changed;
	};

	return {
		subscribe,

		startWithGrid(grid) {
			const game = createGame({ sudoku: createSudoku(grid) });
			set(snapshotGame(game));
		},

		startNew(difficulty) {
			this.startWithGrid(generateSudoku(difficulty));
		},

		startCustom(sencode) {
			this.startWithGrid(decodeSencode(sencode));
		},

		guess(pos, value) {
			if (!pos || pos.x == null || pos.y == null) {
				return false;
			}

			return mutateGame((game) => game.guess({ row: pos.y, col: pos.x, value }));
		},

		applyHint(pos) {
			if (!pos || pos.x == null || pos.y == null) {
				return false;
			}

			return mutateGame((game) => {
				const solvedGrid = solveSudoku(game.getSudoku().getGrid());
				return game.guess({ row: pos.y, col: pos.x, value: solvedGrid[pos.y][pos.x] });
			});
		},

		undo() {
			return mutateGame((game) => game.undo());
		},

		redo() {
			return mutateGame((game) => game.redo());
		},
	};
}

export const gameSession = createGameSession();
